src/compression/huffman.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::{BTreeMap, BinaryHeap, HashMap};
+
+#[derive(Eq, PartialEq, Debug, Ord, PartialOrd)]
+pub enum HuffmanNode {
+    Internal {
+        frequency: usize,
+        left: Box<HuffmanNode>,
+        right: Box<HuffmanNode>,
+    },
+    Leaf {
+        character: char,
+        frequency: usize,
+    },
+}
+
+pub fn huffman_encode(text: &str) -> (String, BTreeMap<char, String>) {
+    let mut frequencies = HashMap::new();
+
+    for char in text.chars() {
+        let counter = frequencies.entry(char).or_insert(0);
+        *counter += 1;
+    }
+
+    let mut heap = BinaryHeap::new();
+    for (char, freq) in frequencies.iter() {
+        heap.push(HuffmanNode::Leaf {
+            character: *char,
+            frequency: *freq,
+        });
+    }
```

Comment:
```suggestion
    let mut heap: BinaryHeap<_> = frequencies
        .iter()
        .map(|(&char, &freq)| HuffmanNode::Leaf {
            character: char,
            frequency: freq,
        })
        .collect();
```

As above: I would suggest a separate function.